src/notebooks/record_dedupe.py

```diff
@@ -0,0 +1,111 @@
+    df.loc[9, 'birthDate'] = '-'.join(dob_list)
+
+    return df[:10]
+
+
+def create_linking_identifier(df: pd.DataFrame):
+    '''
+    Creates a basic identifier used to check duplicates in an incoming
+    batch of data. The basic identifier for the prototype is
+      FIRSTNAME-LASTNAME-YYYY-MM-DD
+    This identifier is joined to the dataframe proper for use in future
+    computation, as well as potential future use in generating/referencing
+    a master ID.
+    '''
+    df['last_name'] = df['name'].apply(lambda x: x[0]['family'] if type(x[0]['family']) is str else '-'.join(x[0]['family']))
+    df['first_name'] = df['name'].apply(lambda x: x[0]['given'] if type(x[0]['given']) is str else '-'.join(list(x[0]['given'])))
+    df['linking_identifier'] = df['first_name'] + '-' + df['last_name'] + '-' + df['birthDate']
+    df.drop(columns=['first_name', 'last_name'], inplace=True)
+    return df
+
+
+def find_matches(df: pd.DataFrame, dupe_field: str):
+    '''
+    From a dataframe containing a column to use as the unique identifier
+    field, identify any and all duplicate records using strict, EXACT
+    match criteria. Returns a dictionary showing the unique identifiers
+    mapping to lists of rows in the DF that all slotted into this group.
+    Currently not required for linkage, but exists a separate util.
+    '''
+    dupes = {}
+    for id, group in df.groupby([dupe_field]):
+        dupes[id] = list(group.index)
+    return dupes
+
+
+def record_combination_func(x: pd.Series):
+    '''
+    Aggregation function applied behind the scenes when performing
+    de-duplicating linkage. Automatically filters for blank, null,
+    and NaN values to facilitate squashing duplicates down into one
+    consolidated record, such that, for any column X:
+      - if records 1, ..., n-1 are empty in X and record n is not,
+        then n(X) is used as a single value
+      - if some number of records 1, ..., j <= n have the same value 
+        in X and all other records are blank in X, then the value
+        of the matching columns 1, ..., j is used as a singleton
+      - if some number of non-empty in X records 1, ..., j <= n have 
+        different values in X, then all values 1(X), ..., j(X) are
+        concatenated into a unique list of values delimited by |
+    '''
+    if x.isnull().values.all():
+        return ''
+    non_nans = [x for x in list(x.astype(str)) if x != '']
+    return '|'.join(set(non_nans)) if len(non_nans) > 1 else non_nans[0]
+
+
+def merge_duplicates(
+        df: pd.DataFrame,
+        dupe_field: str,
+        group_func: Callable
+    ):
+    '''
+    Perform de-duplication linkage on a dataframe of patient records. 
+    All records that have an *EXACT* match in dupe_field are linked
+    according to the provided group_func. Grouped records are consolidated
+    into single rows, and the de-duplicated dataframe is returned.
+    '''
+    df2 = df.groupby(dupe_field).agg(group_func).reset_index()
```

Comment:
:panda_face: :shrug:

Reply:
If we are going to eliminate duplicate records like this then we must update the patient IDs in all resource that link back to the patient resources we are removing.

Reply:
Dan, yep this is now handled (see above). Graham, lol I too was like I guess 🐼 . Took me like an hour of staring it at the other day to convince myself this was actually the perfect function setup for enforcing exact matching. Groupby returns an iterable with two fields `id, group`, where the `id` parameter is the specific `linking identifier` shared by one or more rows, and `group` is a helpfully named `group` object (basically a Series) that contains views to all the rows defined by `id`. Then, the `.agg` function aggregates all these distinct series into an actual dataframe by doing some string manipulations on their values, throwing out the iterable `id` field, and stacking them on top of each other to produce normal data frame rows.